shade-local/shared/src/main/scala/shade/local/immutable/TimeBasedCache.scala

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright (c) 2012-2017 by its authors. Some rights reserved.
+ * See the project homepage at: https://github.com/monix/shade
+ *
+ * Licensed under the MIT License (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy
+ * of the License at:
+ *
+ * https://github.com/monix/shade/blob/master/LICENSE.txt
+ */
+
+package shade.local.immutable
+
+import shade.local.immutable.TimeBasedCache.{Timestamp, Value}
+import scala.annotation.tailrec
+import scala.collection.immutable.SortedMap
+import scala.concurrent.duration._
+
+/** Describes an immutable cache data-structure.
```

Comment:
The immutable cache thing would be useful in combination with streaming:

```scala
observable.scan((cache, elem) => cache.set(elem.key, elem.value))
```

People working with `Observable` have done this many times, aggregating data in simple `Map` references. Adding expiry logic for those values can also be very useful.